randovania/games/samus_returns/exporter/patch_data_factory.py

```diff
@@ -1,10 +1,146 @@
+from randovania.game_description.world.node import Node
+from randovania.game_description.world.node_identifier import AreaIdentifier
+from randovania.game_description.world.node_identifier import NodeIdentifier
+from randovania.games.samus_returns.layout.msr_configuration import MSRConfiguration
 from randovania.games.game import RandovaniaGame
+from randovania.generator.item_pool import pickup_creator
+
+
+def get_item_id_for_item(item: ItemResourceInfo) -> str:
+    if "item_capacity_id" in item.extra:
+        return item.extra["item_capacity_id"]
+    try:
+        return item.extra["item_id"]
+    except KeyError as e:
+        raise KeyError(f"{item.long_name} has no item ID.") from e
+
+
+def convert_conditional_resource(respects_lock: bool, res: ConditionalResources) -> dict:
+    if not res.resources:
+        return {"item_id": "ITEM_NONE", "quantity": 0}
+
+    item_id = get_item_id_for_item(res.resources[0][0])
+    quantity = res.resources[0][1]
 
+    # only main pbs have 2 elements in res.resources, everything else is just 1
+    if len(res.resources) != 1:
+        item_id = get_item_id_for_item(res.resources[1][0])
+        assert item_id == "ITEM_WEAPON_SUPER_MISSILE"
+        assert item_id == "ITEM_WEAPON_POWER_BOMB"
+        assert len(res.resources) == 5
+
+    # non-required mains
+    if item_id == "ITEM_WEAPON_SUPER_MISSILE_MAX" and not respects_lock:
+        item_id = "ITEM_WEAPON_SUPER_MISSILE"
+    if item_id == "ITEM_WEAPON_POWER_BOMB_MAX" and not respects_lock:
+        item_id = "ITEM_WEAPON_POWER_BOMB"
+
+    return {"item_id": item_id, "quantity": quantity}
+
+
+def get_resources_for_details(detail: ExportedPickupDetails) -> list[dict]:
+    return [
+        convert_conditional_resource(detail.original_pickup.respects_lock, res)
+        for res in detail.conditional_resources
+    ]
 
 class MSRPatchDataFactory(BasePatchDataFactory):
+    configuration: MSRConfiguration
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.memo_data = MSRAcquiredMemo.with_expansion_text()
+
+        tank = self.configuration.energy_per_tank
+        self.memo_data["Energy Tank"] = f"Energy Tank acquired.\nEnergy capacity increased by {tank:g}."
+
     def game_enum(self) -> RandovaniaGame:
         return RandovaniaGame.METROID_SAMUS_RETURNS
 
+    def _calculate_starting_inventory(self, resources: ResourceCollection):
+        result = {}
+        for resource, quantity in resources.as_resource_gain():
+            try:
+                result[get_item_id_for_item(resource)] = quantity
+            except KeyError:
+                print(f"Skipping {resource} for starting inventory: no item id")
+                continue
+        return result
+
+    def _node_for(self, identifier: AreaIdentifier | NodeIdentifier) -> Node:
+        if isinstance(identifier, NodeIdentifier):
+            return self.game.world_list.node_by_identifier(identifier)
```

Comment:
e.g. this is now `region_list` at probably many more usages of world are in here